monit.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.35
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MONIT 1"
.TH MONIT 1 "www.mmonit.com" "July 26. 2011" "User Commands"
.SH "NAME"
Monit \- utility for monitoring services on a Unix system
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBmonit\fR [options] {arguments}
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBmonit\fR is a utility for managing and monitoring processes,
files, directories and filesystems on a Unix system. Monit
conducts automatic maintenance and repair and can execute
meaningful causal actions in error situations. E.g. Monit can
start a process if it does not run, restart a process if it does
not respond and stop a process if it uses too much resources. You
may use Monit to monitor files, directories and filesystems for
changes, such as timestamps changes, checksum changes or size
changes.
.PP
Monit is controlled via an easy to configure control file based
on a free\-format, token-oriented syntax. Monit logs to syslog or
to its own log file and notifies you about error conditions via
customizable alert messages. Monit can perform various \s-1TCP/IP\s0
network checks, protocol checks and can utilize \s-1SSL\s0 for such
checks. Monit provides a http(s) interface and you may use a
browser to access the Monit program.
.SH "GENERAL OPERATION"
.IX Header "GENERAL OPERATION"
The behavior of Monit is controlled by command-line options
\&\fIand\fR a run control file, \fI~/.monitrc\fR, the syntax of which we
describe in a later section. Command-line options override
\&\fI.monitrc\fR declarations.
.PP
The following options are recognized by monit. However, it is
recommended that you set options (when applicable) directly in
the \fI.monitrc\fR control file.
.Sh "General Options and Arguments"
.IX Subsection "General Options and Arguments"
\&\fB\-c\fR \fIfile\fR
   Use this control file
.PP
\&\fB\-d\fR \fIn\fR
   Run as a daemon once per \fIn\fR seconds
.PP
\&\fB\-g\fR 
   Set group name for start, stop, restart, monitor and
   unmonitor.
.PP
\&\fB\-l\fR \fIlogfile\fR
   Print log information to this file 
.PP
\&\fB\-p\fR \fIpidfile\fR
   Use this lock file in daemon mode
.PP
\&\fB\-s\fR \fIstatefile\fR
   Write state information to this file
.PP
\&\fB\-I\fR
   Do not run in background (needed for run from init)
.PP
\&\fB\-t\fR
   Run syntax check for the control file
.PP
\&\fB\-v\fR
   Verbose mode, work noisy (diagnostic output)
.PP
\&\fB\-vv\fR
   Very verbose mode, same as \-v plus log stacktrace on error
.PP
\&\fB\-H\fR \fI[filename]\fR
   Print \s-1MD5\s0 and \s-1SHA1\s0 hashes of the file or of stdin if the 
   filename is omitted; Monit will exit afterwards
.PP
\&\fB\-V\fR
   Print version number and patch level
.PP
\&\fB\-h\fR
   Print a help text
.PP
In addition to the options above, Monit can be started with one
of the following action arguments; Monit will then execute the
action and exit without transforming itself to a daemon.
.PP
\&\fBstart all\fR
   Start all services listed in the control file and 
   enable monitoring for them. If the group option is 
   set, only start and enable monitoring of services in
   the named group (no \*(L"all\*(R" verb is required in this
   case).
.PP
\&\fBstart name\fR
   Start the named service and enable monitoring for 
   it. The name is a service entry name from the 
   monitrc file.
.PP
\&\fBstop all\fR     
   Stop all services listed in the control file and 
   disable their monitoring. If the group option is 
   set, only stop and disable monitoring of the services 
   in the named group (no \*(L"all\*(R" verb is required in this
   case).
.PP
\&\fBstop name\fR
   Stop the named service and disable its monitoring. 
   The name is a service entry name from the monitrc 
   file. 
.PP
\&\fBrestart all\fR
   Stop and start \fIall\fR services. If the group option 
   is set, only restart the services in the named group
   (no \*(L"all\*(R" verb is required in this case).
.PP
\&\fBrestart name\fR
   Restart the named service. The name is a service entry 
   name from the monitrc file.
.PP
\&\fBmonitor all\fR     
   Enable monitoring of all services listed in the
   control file. If the group option is set, only start
   monitoring of services in the named group (no \*(L"all\*(R"
   verb is required in this case).
.PP
\&\fBmonitor name\fR

CHANGES
                     CHANGES version 5.2.7
                                
           This file summarizes changes made since 5.0
                                
       Bug fixes may refer to an auto-generated id in our
          bug-tracking system, which can be found here
          https://savannah.nongnu.org/bugs/?group=monit


Version 5.2.7

* Added set exec to apply when alert is set out

Version 5.2.6

* Fix MySQL protocol test: MySQL 5.5.12 returns new error code in
  the case of authentication failure.

* Fix Debian bug #621047: monit fails to build after SSLv2 removal

* Fix crash on Solaris which may occur if the system load is zero.
  Thanks to Paul Sun for report.

* The stacktrace logging on error is disabled in -v (verbose) mode
  as it was too verbose for common service debugging tasks, it can
  be enabled using -vv option.

* Improve how fast Monit check if a program was started or stopped.
  Thanks to Michael Renner for patch. 



Version 5.2.5

* Fix process match check - when the monitored process failed and
  was restarted by Monit, Monit didn't recognized  it is  running
  after the restart and reported  start failure (similar on stop).
  Thanks to  Kenichi Futatsumori for report and  helping to root
  cause the problem.

* Fix Debian #617259: symbolic links in the filesystem check doesn't
  work.  Thanks to Sergey B Kirpichev for report.

* Fix Debian bug #614984: smtp protocol test issues both EHLO and
  HELO.  Thanks to Sergey B Kirpichev for report.

* Fix bug #32583: Multiple SIP OPTIONS messages use the same header
  data.  Thanks to Hugh Waite for patch.

* Try harder to get FQDN hostname for the host where monit is running.
  The hostname in the $HOST variable which is used in the mail sender
  may thus change. Thanks to Sergey B Kirpichev for patch.

* AIX: Fix the time display which was off by GMT difference. Thanks
  to Helen Chen for report.

* AIX: Fix the M/Monit heartbeat. Thanks to Helen Chen for report.

* Support symbolic link to monit configuration file.

* Fix crash when monit daemon start delay option was used and monit
  was signalized to stop before the start delay passed. Thanks to
  John Schult for report.



Version 5.2.4

NEW FEATURES AND FUNCTIONS:

* Added the "procmatch" CLI command which allows for easy testing
  of pattern for process match check. The command takes regular
  expression as an argument and displays all running processes
  matching the pattern. Example usage:
      $ monit procmatch "iChatAgent"

* Set the default log file mask to 0640 (originally it was 0664).
  Thanks to Sergey B Kirpichev.

* Reduced monit memory footprint by ca. 10%.

BUGFIXES:

* FreeBSD, NetBSD, OpenBSD, MacOSX, Solaris filesystem check fix:
  If block/character device was used in the filesystem path instead
  of mountpoint, monit reported usage of wrong filesystem.

* NetBSD filesystem check: Fix space usage report.

* Fix memory usage monitoring in OpenVZ VPS 2.6.32 virtual hosts.
  Thanks to Kelly for report.

* If the protocol test failed, show the request in the event. Thanks
  to Marco for report.

* Randomize the mail message id to prevent duplicates in the case, that
  the same hostname is used on multiple hosts running monit and messages
  are generated in the same second in parallel. Thanks to Sergey B
  Kirpichev.

* Spelling fixes. Thanks to Sergey B Kirpichev.



Version 5.2.3

BUGFIXES:

* Mysql protocol test supports mysql 5.5.x and newer now.



Version 5.2.2

BUGFIXES:

* Fix crash on MacOSX

* ICMP echo test (ping):
     - bug #31128: do not log error if different response type is received
     - bug #31129: do not require root to use ping test. Privilege to create
       raw socket is still required, but on some platforms such as Solaris it
       can be granted to non-root users too. If the user has no permission to
       perform ping, monit will skip the icmp test and log message (in debug
       mode only).


src/monit.c
/*
 * Copyright (C) Tildeslash Ltd. All rights reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * In addition, as a special exception, the copyright holders give
 * permission to link the code of portions of this program with the
 * OpenSSL library under certain conditions as described in each
 * individual source file, and distribute linked combinations
 * including the two.
 *
 * You must obey the GNU Affero General Public License in all respects
 * for all of the code used other than OpenSSL.  
 */


#include "config.h"
#include <locale.h>

#ifdef HAVE_STDIO_H
#include <stdio.h>
#endif

#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif

#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif

#ifdef HAVE_SIGNAL_H
#include <signal.h>
#endif

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#ifdef HAVE_STRING_H
#include <string.h>
#endif

#ifdef HAVE_STRINGS_H
#include <strings.h>
#endif

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif

#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif


#include "monit.h"
#include "net.h"
#include "ssl.h"
#include "process.h"
#include "md5.h"
#include "sha.h"
#include "state.h"
#include "event.h"


/**
 *  DESCRIPTION
 *    monit - system for monitoring services on a Unix system
 *
 *  SYNOPSIS
 *    monit [options] {arguments}
 *
 *  @author Jan-Henrik Haukeland, <hauk@tildeslash.com>
 *  @author Martin Pala <martinp@tildeslash.com>
 *  @author Christian Hopp, <chopp@iei.tu-clausthal.de>
 *
 *  @file
 */


/* -------------------------------------------------------------- Prototypes */


static void  do_init();                       /* Initialize this application */
static void  do_reinit();           /* Re-initialize the runtime application */
static void  do_action(char **);         /* Dispatch to the submitted action */
static void  do_exit();                                    /* Finalize monit */
static void  do_default();                              /* Do default action */
static void  handle_options(int, char **);         /* Handle program options */
static void  help();                 /* Print program help message to stdout */
static void  version();                         /* Print version information */
static void *heartbeat(void *args);              /* M/Monit heartbeat thread */
static RETSIGTYPE do_reload(int);       /* Signalhandler for a daemon reload */
static RETSIGTYPE do_destroy(int);   /* Signalhandler for monit finalization */
static RETSIGTYPE do_wakeup(int);  /* Signalhandler for a daemon wakeup call */


/* ------------------------------------------------------------------ Global */


char   *prog;                                  /**< The Name of this Program */
struct myrun Run;                      /**< Struct holding runtime constants */
Service_T servicelist;                /**< The service list (created in p.y) */
Service_T servicelist_conf;   /**< The service list in conf file (c. in p.y) */
ServiceGroup_T servicegrouplist;/**< The service group list (created in p.y) */
SystemInfo_T systeminfo;                              /**< System infomation */

pthread_t           heartbeatThread;           /**< M/Monit heartbeat thread */
pthread_cond_t      heartbeatCond;            /**< Hearbeat wakeup condition */
pthread_mutex_t     heartbeatMutex;                      /**< Hearbeat mutex */
static volatile int heartbeatRunning = FALSE;     /**< Heartbeat thread flag */

int ptreesize = 0;
int oldptreesize = 0;
ProcessTree_T *ptree = NULL;
ProcessTree_T *oldptree = NULL;

char actionnames[][STRLEN]   = {"ignore", "alert", "restart", "stop", "exec", "unmonitor", "start", "monitor", ""};
char modenames[][STRLEN]     = {"active", "passive", "manual"};
char checksumnames[][STRLEN] = {"UNKNOWN", "MD5", "SHA1"};
char operatornames[][STRLEN] = {"greater than", "less than", "equal to", "not equal to"};
char operatorshortnames[][3] = {">", "<", "=", "!="};
char monitornames[][STRLEN]  = {"not monitored", "monitored", "initializing"};
char statusnames[][STRLEN]   = {"accessible", "accessible", "accessible", "running", "online with all services", "running", "accessible"};
char servicetypes[][STRLEN]  = {"Filesystem", "Directory", "File", "Process", "Remote Host", "System", "Fifo"};
char pathnames[][STRLEN]     = {"Path", "Path", "Path", "Pid file", "Path", "", "Path"};
char icmpnames[19][STRLEN]   = {"Echo Reply", "", "", "Destination Unreachable", "Source Quench", "Redirect", "", "", "Echo Request", "", "", "Time Exceeded", "Parameter Problem", "Timestamp Request", "Timestamp Reply", "Information Request", "Information Reply", "Address Mask Request", "Address Mask Reply"};
char sslnames[][STRLEN]      = {"auto", "v2", "v3", "tls"};




/* ------------------------------------------------------------------ Public */


/**
 * The Prime mover
 */
int main(int argc, char **argv) {
  setlocale(LC_ALL, "C");
  prog = Util_basename(argv[0]);
  init_env();
  handle_options(argc, argv);
 
  do_init();
  do_action(argv); 
  do_exit();

  return 0;
}


/**
 * Wakeup a sleeping monit daemon.
 * Returns TRUE on success otherwise FALSE
 */
int do_wakeupcall() {
  pid_t pid;
  
  if ((pid = exist_daemon()) > 0) {
    kill(pid, SIGUSR1);
    LogInfo("%s daemon at %d awakened\n", prog, pid);
    
    return TRUE;
  }
  
  return FALSE;
}


/* ----------------------------------------------------------------- Private */


/**
 * Initialize this application - Register signal handlers,
 * Parse the control file and initialize the program's
 * datastructures and the log system.
 */
static void do_init() {

  int status;

  /*
   * Register interest for the SIGTERM signal,
   * in case we run in daemon mode this signal
   * will terminate a running daemon.
   */
  signal(SIGTERM, do_destroy);

  /*
   * Register interest for the SIGUSER1 signal,
   * in case we run in daemon mode this signal
   * will wakeup a sleeping daemon.
   */
  signal(SIGUSR1, do_wakeup);

  /*
   * Register interest for the SIGINT signal,
   * in case we run as a server but not as a daemon
   * we need to catch this signal if the user pressed
   * CTRL^C in the terminal
   */
  signal(SIGINT, do_destroy);

  /*
   * Register interest for the SIGHUP signal,
   * in case we run in daemon mode this signal
   * will reload the configuration.
   */
  signal(SIGHUP, do_reload);

  /*
   * Register no interest for the SIGPIPE signal,
   */
  signal(SIGPIPE, SIG_IGN);

  /*
   * Initialize the random number generator
   */
  srandom(time(NULL) + getpid());

  /*
   * Initialize the Runtime mutex. This mutex
   * is used to synchronize handling of global
   * service data
   */
  status = pthread_mutex_init(&Run.mutex, NULL);
  if (status != 0) {
    LogError("%s: Cannot initialize mutex -- %s\n", prog, strerror(status));
    exit(1);
  }

  /*
   * Initialize heartbeat mutex and condition
   */
  status = pthread_mutex_init(&heartbeatMutex, NULL);
  if (status != 0) {
    LogError("%s: Cannot initialize heartbeat mutex -- %s\n", prog, strerror(status));
    exit(1);
  }
  status = pthread_cond_init(&heartbeatCond, NULL);
  if (status != 0) {
    LogError("%s: Cannot initialize heartbeat condition -- %s\n", prog, strerror(status));
    exit(1);
  }

  /* 
   * Get the position of the control file 
   */
  if (! Run.controlfile)
    Run.controlfile = file_findControlFile();
  
  /*
   * Initialize the process information gathering interface
   */
  Run.doprocess = init_process_info();

  /*
   * Start the Parser and create the service list. This will also set
   * any Runtime constants defined in the controlfile.
   */
  if (! parse(Run.controlfile))
    exit(1);

  /*
   * Stop and report success if we are just validating the Control
   * file syntax. The previous parse statement exits the program with
   * an error message if a syntax error is present in the control
   * file.
   */
  if (Run.testing) {
    LogInfo("Control file syntax OK\n");
    exit(0);
  }

  /*
   * Initialize the log system 
   */
  if (! log_init())
    exit(1);

  /* 
   * Did we find any service ?  
   */
  if (! servicelist) {
    LogError("%s: No services has been specified\n", prog);
    exit(0);
  }
  
  /* 
   * Initialize Runtime file variables 
   */
  file_init();

  /* 
   * Should we print debug information ? 
   */
  if (Run.debug) {
    Util_printRunList();
    Util_printServiceList();
  }
}


/**
 * Re-Initialize the application - called if a
 * monit daemon receives the SIGHUP signal.
 */
static void do_reinit() {
  int status;

  LogInfo("Awakened by the SIGHUP signal\n");
  LogInfo("Reinitializing %s - Control file '%s'\n", prog, Run.controlfile);
  
  if(Run.mmonits && heartbeatRunning) {
    if ((status = pthread_cond_signal(&heartbeatCond)) != 0)
      LogError("%s: Failed to signal the heartbeat thread -- %s\n", prog, strerror(status));
    if ((status = pthread_join(heartbeatThread, NULL)) != 0)
      LogError("%s: Failed to stop the heartbeat thread -- %s\n", prog, strerror(status));
    heartbeatRunning = FALSE;
  }

  Run.doreload = FALSE;
  
  /* Stop http interface */
  if (Run.dohttpd)
    monit_http(STOP_HTTP);

  /* Save the current state (no changes are possible now
     since the http thread is stopped) */
  State_save();

  /* Run the garbage collector */
  gc();

  if (! parse(Run.controlfile)) {
    LogError("%s daemon died\n", prog);
    exit(1);
  }

  /* Close the current log */
  log_close();

  /* Reinstall the log system */
  if (! log_init())
    exit(1);

  /* Did we find any services ?  */
  if (! servicelist) {
    LogError("%s: No services has been specified\n", prog);
    exit(0);
  }
  
  /* Reinitialize Runtime file variables */
  file_init();

  if (! file_createPidFile(Run.pidfile)) {
    LogError("%s daemon died\n", prog);
    exit(1);
  }

  /* Update service data from the state repository */
  State_update();
  
  /* Start http interface */
  if (can_http())
    monit_http(START_HTTP);

  /* send the monit startup notification */
  Event_post(Run.system, Event_Instance, STATE_CHANGED, Run.system->action_MONIT_RELOAD, "Monit reloaded");

  if(Run.mmonits && ((status = pthread_create(&heartbeatThread, NULL, heartbeat, NULL)) != 0))
    LogError("%s: Failed to create the heartbeat thread -- %s\n", prog, strerror(status));
  else
    heartbeatRunning = TRUE;
}


/**
 * Dispatch to the submitted action - actions are program arguments
 */
static void do_action(char **args) {
  char *action = args[optind];
  char *service = args[++optind];

  Run.once = TRUE;

  if (! action) {
    do_default();
  } else if (IS(action, "start")     ||
             IS(action, "stop")      ||
             IS(action, "monitor")   ||
             IS(action, "unmonitor") ||
             IS(action, "restart")) {
    if (Run.mygroup || service) {
      int errors = 0;
      int (*_control_service)(const char *, const char *) = exist_daemon() ? control_service_daemon : control_service_string;

      if (Run.mygroup) {
        ServiceGroup_T sg = NULL;

        for (sg = servicegrouplist; sg; sg = sg->next) {
          if (! strcasecmp(Run.mygroup, sg->name)) {
            ServiceGroupMember_T sgm = NULL;

            for (sgm = sg->members; sgm; sgm = sgm->next)
              if (! _control_service(sgm->name, action))
                errors++;

            break;
          }
        }
      } else if (IS(service, "all")) {
        Service_T s = NULL;

        for (s = servicelist; s; s = s->next) {
          if (s->visited)
            continue;
          if (! _control_service(s->name, action))
              errors++;
        }
      } else {
        errors = _control_service(service, action) ? 0 : 1;
      }
      if (errors)
        exit(1);
    } else {
      LogError("%s: please specify the configured service name or 'all' after %s\n", prog, action);
      exit(1);
    }
  } else if (IS(action, "reload")) {
    LogInfo("Reinitializing monit daemon\n", prog);
    kill_daemon(SIGHUP);
  } else if (IS(action, "status")) {
    status(LEVEL_NAME_FULL);
  } else if (IS(action, "summary")) {
    status(LEVEL_NAME_SUMMARY);
  } else if (IS(action, "procmatch")) {
    if (! service) {
      printf("Invalid syntax - usage: procmatch \"<pattern>\"\n");
      exit(1);
    }
    process_testmatch(service);
  } else if (IS(action, "quit")) {
    kill_daemon(SIGTERM);
  } else if (IS(action, "validate")) {
    if (! validate())
      exit(1);
  } else {
    LogError("%s: invalid argument -- %s  (-h will show valid arguments)\n", prog, action);
    exit(1);
  }
}


/**
 * Finalize monit
 */
static void do_exit() {
  int status;
  sigset_t ns;

  set_signal_block(&ns, NULL);
  Run.stopped = TRUE;
  if (Run.isdaemon && !Run.once) {
    if (can_http())
      monit_http(STOP_HTTP);

    if(Run.mmonits && heartbeatRunning) {
      if ((status = pthread_cond_signal(&heartbeatCond)) != 0)
        LogError("%s: Failed to signal the heartbeat thread -- %s\n", prog, strerror(status));
      if ((status = pthread_join(heartbeatThread, NULL)) != 0)
        LogError("%s: Failed to stop the heartbeat thread -- %s\n", prog, strerror(status));
      heartbeatRunning = FALSE;
    }

    LogInfo("%s daemon with pid [%d] killed\n", prog, (int)getpid());

    /* send the monit stop notification */
    Event_post(Run.system, Event_Instance, STATE_CHANGED, Run.system->action_MONIT_STOP, "Monit stopped");
  }
  gc();
  exit(0);
}


/**
 * Default action - become a daemon if defined in the Run object and
 * run validate() between sleeps. If not, just run validate() once.
 * Also, if specified, start the monit http server if in deamon mode.
 */
static void do_default() {
  int status;

  if (Run.isdaemon) {
    if (do_wakeupcall())
      exit(0);
  
    Run.once = FALSE;
    if (can_http())
      LogInfo("Starting %s daemon with http interface at [%s:%d]\n", prog, Run.bind_addr?Run.bind_addr:"*", Run.httpdport);
    else
      LogInfo("Starting %s daemon\n", prog);
    
    if (Run.startdelay)
        LogInfo("Monit start delay set -- pause for %ds\n", Run.startdelay);

    if (Run.init != TRUE)
      daemonize(); 
    else if (! Run.debug)
      Util_redirectStdFds();
    
    if (! file_createPidFile(Run.pidfile)) {
      LogError("%s daemon died\n", prog);
      exit(1);
    }

    if (State_shouldUpdate())
      State_update();

    atexit(file_finalize);
  
    if (Run.startdelay) {
	time_t now = time(NULL);
        time_t delay = now + Run.startdelay;

        /* sleep can be interrupted by signal => make sure we paused long enough */
        while (now < delay) {
    	  sleep(delay - now);
          if (Run.stopped)
            do_exit();
          now = time(NULL);
        }
    }

    if (can_http())
      monit_http(START_HTTP);
    
    /* send the monit startup notification */
    Event_post(Run.system, Event_Instance, STATE_CHANGED, Run.system->action_MONIT_START, "Monit started");

    if(Run.mmonits && ((status = pthread_create(&heartbeatThread, NULL, heartbeat, NULL)) != 0))
      LogError("%s: Failed to create the heartbeat thread -- %s\n", prog, strerror(status));
    else
      heartbeatRunning = TRUE;

    while (TRUE) {
      validate();
      State_save();

      /* In the case that there is no pending action then sleep */
      if (!Run.doaction)
        sleep(Run.polltime);

      if (Run.dowakeup) {
        Run.dowakeup = FALSE;
        LogInfo("Awakened by User defined signal 1\n");
      }
      
      if (Run.stopped)
        do_exit();
      else if (Run.doreload)
        do_reinit();
    }
  } else {
    validate();
  }
}


/**
 * Handle program options - Options set from the commandline
 * takes precedence over those found in the control file
 */
static void handle_options(int argc, char **argv) {
  int opt;
  opterr = 0;
  Run.mygroup = NULL;

  while ((opt = getopt(argc,argv,"c:d:g:l:p:s:iItvVhH")) != -1) {

    switch(opt) {

    case 'c':
        Run.controlfile = xstrdup(optarg);
        break;
	
    case 'd':
	Run.isdaemon = TRUE;
 	sscanf(optarg, "%d", &Run.polltime);
	if (Run.polltime<1) {
	  LogError("%s: option -%c requires a natural number\n", prog, opt);
	  exit(1);
	}
	break;

    case 'g':
        Run.mygroup = xstrdup(optarg);
        break;
	
    case 'l':
        Run.logfile = xstrdup(optarg);
	if (IS(Run.logfile, "syslog"))
	    Run.use_syslog = TRUE;
	Run.dolog = TRUE;
        break;
   
    case 'p':
        Run.pidfile = xstrdup(optarg);
        break;

    case 's':
        Run.statefile = xstrdup(optarg);
        break;

    case 'I':
	Run.init = TRUE;
	break;
      
    case 't':
        Run.testing = TRUE;
        break;
	
    case 'v':
        Run.debug++;
        break;

    case 'H':
        if (argc > optind)
          Util_printHash(argv[optind]);
        else
          Util_printHash(NULL);
          
        exit(0);
	break;
	
    case 'V':
        version();
        exit(0);
	break;
	
    case 'h':
        help();
        exit(0);
	break;
	
    case '?':
	switch(optopt) {
	  
	case 'c':
	case 'd':
	case 'g':
	case 'l':
	case 'p':
	case 's':
	    LogError("%s: option -- %c requires an argument\n", prog, optopt);
	    break;
	default:
	    LogError("%s: invalid option -- %c  (-h will show valid options)\n", prog, optopt);
	    
	}
	
	exit(1);
	
    }
    
  }
  
}


/**
 * Print the program's help message
 */
static void help() {
  printf("Usage: %s [options] {arguments}\n", prog);
  printf("Options are as follows:\n");
  printf(" -c file       Use this control file\n");
  printf(" -d n          Run as a daemon once per n seconds\n");
  printf(" -g name       Set group name for start, stop, restart, monitor and unmonitor\n");
  printf(" -l logfile    Print log information to this file\n");
  printf(" -p pidfile    Use this lock file in daemon mode\n");
  printf(" -s statefile  Set the file monit should write state information to\n");
  printf(" -I            Do not run in background (needed for run from init)\n");
  printf(" -t            Run syntax check for the control file\n");
  printf(" -v            Verbose mode, work noisy (diagnostic output)\n");
  printf(" -vv           Very verbose mode, same as -v plus log stacktrace on error\n");
  printf(" -H [filename] Print SHA1 and MD5 hashes of the file or of stdin if the\n");
  printf("               filename is omited; monit will exit afterwards\n");
  printf(" -V            Print version number and patchlevel\n");
  printf(" -h            Print this text\n");
  printf("Optional action arguments for non-daemon mode are as follows:\n");
  printf(" start all           - Start all services\n");
  printf(" start name          - Only start the named service\n");
  printf(" stop all            - Stop all services\n");
  printf(" stop name           - Only stop the named service\n");
  printf(" restart all         - Stop and start all services\n");
  printf(" restart name        - Only restart the named service\n");
  printf(" monitor all         - Enable monitoring of all services\n");
  printf(" monitor name        - Only enable monitoring of the named service\n");
  printf(" unmonitor all       - Disable monitoring of all services\n");
  printf(" unmonitor name      - Only disable monitoring of the named service\n");
  printf(" reload              - Reinitialize monit\n");
  printf(" status              - Print full status information for each service\n");
  printf(" summary             - Print short status information for each service\n");
  printf(" quit                - Kill monit daemon process\n");
  printf(" validate            - Check all services and start if not running\n");
  printf(" procmatch <pattern> - Test process matching pattern\n");
  printf("\n");
  printf("(Action arguments operate on services defined in the control file)\n");
}

/**
 * Print version information
 */
static void version() {
  printf("This is Monit version " VERSION "\n");
  printf("Copyright (C) 2000-2011 Tildeslash Ltd.");
  printf(" All Rights Reserved.\n");
}


/**
 * M/Monit heartbeat thread
 */
static void *heartbeat(void *args) {
  sigset_t ns;
  struct timespec wait;

  set_signal_block(&ns, NULL);
  LogInfo("M/Monit heartbeat started\n");
  LOCK(heartbeatMutex)
  {
    while (! Run.stopped && ! Run.doreload) {
      if (handle_mmonit(NULL) == HANDLER_SUCCEEDED)
        wait.tv_sec = time(NULL) + Run.polltime;
      else
        wait.tv_sec = time(NULL) + 1;
      wait.tv_nsec = 0;
      pthread_cond_timedwait(&heartbeatCond, &heartbeatMutex, &wait);
    }
  }
  END_LOCK;
  LogInfo("M/Monit heartbeat stopped\n");
  return NULL;
}


/**
 * Signalhandler for a daemon reload call
 */
static RETSIGTYPE do_reload(int sig) {
  Run.doreload = TRUE;
}


/**
 * Signalhandler for monit finalization
 */
static RETSIGTYPE do_destroy(int sig) {
  Run.stopped = TRUE;
}


/**
 * Signalhandler for a daemon wakeup call
 */
static RETSIGTYPE do_wakeup(int sig) {
  Run.dowakeup = TRUE;
}

